// @eventstore/db-client/dist/persistentSubscription/utils/persistentSubscriptionSettings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.persistentSubscriptionToAllSettingsFromDefaults = exports.persistentSubscriptionToStreamSettingsFromDefaults = void 0;
const constants_1 = require("../../constants");
const defaults = {
    resolveLinkTos: false,
    extraStatistics: false,
    messageTimeout: 30000,
    maxRetryCount: 10,
    checkPointAfter: 2000,
    checkPointLowerBound: 10,
    checkPointUpperBound: 1000,
    maxSubscriberCount: constants_1.UNBOUNDED,
    liveBufferSize: 500,
    readBatchSize: 20,
    historyBufferSize: 500,
    consumerStrategyName: constants_1.ROUND_ROBIN,
};
/**
 * Creates {@link PersistentSubscriptionToStreamSettings} from default settings.
 * @param changes Changes to apply to the default settings.
 */
const persistentSubscriptionToStreamSettingsFromDefaults = (changes = {}) => ({
    startFrom: constants_1.END,
    ...defaults,
    ...changes,
});
exports.persistentSubscriptionToStreamSettingsFromDefaults = persistentSubscriptionToStreamSettingsFromDefaults;
/**
 * Creates {@link PersistentSubscriptionToAllSettings} from default settings.
 * @param changes Changes to apply to the default settings.
 */
const persistentSubscriptionToAllSettingsFromDefaults = (changes = {}) => ({
    startFrom: constants_1.END,
    ...defaults,
    ...changes,
});
exports.persistentSubscriptionToAllSettingsFromDefaults = persistentSubscriptionToAllSettingsFromDefaults;
