// @eventstore/db-client/dist/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SYSTEM_STREAM_ACL = exports.USER_STREAM_ACL = exports.STATS_STREAM_PREFIX = exports.SETTINGS_STREAM = exports.STREAMS_STREAM = exports.UNBOUNDED = exports.EVENT_TYPE = exports.STREAM_NAME = exports.WRITING_RESULTS = exports.PAUSED = exports.DELETING = exports.FAULTED = exports.ABORTED = exports.COMPLETED = exports.STOPPED = exports.ABORTING = exports.STOPPING = exports.RUNNING = exports.LOADING_STOPPED = exports.STARTING = exports.PREPARED = exports.PREPARING = exports.LOADED = exports.LOADING = exports.CREATING = exports.STOP = exports.SKIP = exports.RETRY = exports.PARK = exports.PINNED = exports.ROUND_ROBIN = exports.DISPATCH_TO_SINGLE = exports.READ_ONLY_REPLICA = exports.LEADER = exports.FOLLOWER = exports.RANDOM = exports.BACKWARDS = exports.FORWARDS = exports.END = exports.START = exports.NO_STREAM = exports.STREAM_EXISTS = exports.ANY = void 0;
// revision
exports.ANY = "any";
exports.STREAM_EXISTS = "stream_exists";
exports.NO_STREAM = "no_stream";
// known positions / revisions
exports.START = "start";
exports.END = "end";
// directions
exports.FORWARDS = "forwards";
exports.BACKWARDS = "backwards";
// node Preference
exports.RANDOM = "random";
exports.FOLLOWER = "follower";
exports.LEADER = "leader";
exports.READ_ONLY_REPLICA = "read_only_replica";
// consumer strategy
exports.DISPATCH_TO_SINGLE = "DispatchToSingle";
exports.ROUND_ROBIN = "RoundRobin";
exports.PINNED = "Pinned";
// persistent action
exports.PARK = "park";
exports.RETRY = "retry";
exports.SKIP = "skip";
exports.STOP = "stop";
// projection status
exports.CREATING = "Creating";
exports.LOADING = "Loading";
exports.LOADED = "Loaded";
exports.PREPARING = "Preparing";
exports.PREPARED = "Prepared";
exports.STARTING = "Starting";
exports.LOADING_STOPPED = "LoadingStopped";
exports.RUNNING = "Running";
exports.STOPPING = "Stopping";
exports.ABORTING = "Aborting";
exports.STOPPED = "Stopped";
exports.COMPLETED = "Completed";
exports.ABORTED = "Aborted";
exports.FAULTED = "Faulted";
exports.DELETING = "Deleting";
// processing status
exports.PAUSED = "Paused";
exports.WRITING_RESULTS = "Writing results";
// STOPPED
// filter
exports.STREAM_NAME = "streamName";
exports.EVENT_TYPE = "eventType";
// max subscriber count
exports.UNBOUNDED = "unbounded";
// system streams
/**
 * A stream containing links pointing to each stream in the EventStoreDB.
 */
exports.STREAMS_STREAM = "$streams";
/**
 * A stream containing system settings.
 */
exports.SETTINGS_STREAM = "$settings";
/**
 * A stream containing statistics.
 */
exports.STATS_STREAM_PREFIX = "$stats";
// acl defaults
/**
 * The user default acl stream.
 */
exports.USER_STREAM_ACL = "$userStreamAcl";
/**
 * The system stream defaults acl stream.
 */
exports.SYSTEM_STREAM_ACL = "$systemStreamAcl";
